### ISSUE TITLE
[Meta] Miscellaneous issues while trying to setup the build environment (on a new M1 computer)

### ISSUE THREAD

**Issue by jonalmeida** [mozilla/application-services#5129]
OS: Mac OS 12.5.1 (21G83)
Hardware: MacBook Pro (16-inch, 2021)

While trying to setup app services on a new machine (for myself and onboarding new hires), we've identified a few paper cuts that would be good to fix. Listing them all here under one issue in a meta, but some might be duplicates of the same issue or a non-issue (it's unclear to me):

* Python is missing:
	```
	$ ./libs/verify-android-environment.sh

	/Users/jalmeida/src/application-services/libs/nss-3.66/nss/build.sh -v --opt --static --disable-tests -Ddisable_dbm=1 -Dsign_libs=0 -Ddisable_libpkix=1
	/Users/jalmeida/src/application-services/libs/nss-3.66/nss/build.sh: line 72: python: command not found
	```
	**Workaround:** `export PATH="$PATH:/opt/homebrew/opt/python@3.9/libexec/bin"` is the solution for now that is already documented, but this is not scalable - with every new version of python, we have to update our environment. Why can we not fix this to use `python3` from the existing path (which homebrew sets up for us too and is used by m-c‘s mach as well)?

* Missing (android) sdkmanager:
	```
	$ ./libs/verify-android-environment.sh

	/Users/jalmeida/src/application-services/libs/verify-android-ci-environment.sh: line 67: /Users/jalmeida/.mozbuild/android-sdk-macosx/cmdline-tools/latest/bin/sdkmanager: No such file or directory
	```
		
	**Workaround:** Android SDK manager is now `android-sdk-macosx/cmdline-tools/[version]/bin/sdkmanager`. Add an alias to latest version that is named `latest`.

* Missing `six` python package:
	```
	$ ./libs/verify-android-environment.sh


	  File "/opt/homebrew/lib/python3.10/site-packages/gyp-0.1-py3.10.egg/gyp/__init__.py", line 10, in <module>
	    import gyp.input
	  File "/opt/homebrew/lib/python3.10/site-packages/gyp-0.1-py3.10.egg/gyp/input.py", line 8, in <module>
	    import gyp.common
	  File "/opt/homebrew/lib/python3.10/site-packages/gyp-0.1-py3.10.egg/gyp/common.py", line 14, in <module>
	    from six.moves import collections_abc
	ModuleNotFoundError: No module named 'six'
	```
		
	**Workaround:** Manually install with `pip3 install six`

* Certain failures are incorrectly reported:
	```
	$ ./libs/verify-android-environment.sh

	/Users/jalmeida/src/application-services/libs/verify-android-ci-environment.sh: line 45: die: command not found
	```
		
	**Workaround:** replace "die" with exit or alternative command; the real error is on the line number it references. So look at the script to see what is the real problem (in this example line 45).

* NDK is missing:
	```
	$ ./libs/verify-android-environment.sh
	
	* What went wrong:
	A problem occurred evaluating project ':autofill'.
	> A problem occurred configuring project ':full-megazord'.
	   > Failed to notify project evaluation listener.
	      > NDK is not installed
	```
	**Workaround:** Set the property `ndk.dir=` in a `local.properties` file. Note: Environment vars for the NDK is deprecated and will stop being supported in gradle eventually.

* After setting up the NDK (to the same version used by m-c), if the version doesn't match, it downloads a new NDK.

	**Workaround:** n/a. The verify script should check and download the NDK when the previous error fails. In my case, I had rev r21d and the script wanted to download a new copy of it again, so it needs to ensure the right version already exists or if needs to be in a particular place.

* Warning about architecture can be inaccurate:
	```
	$ ./libs/verify-android-environment.sh

	WARNING: mismatching Java (with
	x86_64]
	arm64e]
	64-bit
	64-bit) and system (arm64) architectures. Make sure this is intentional, or you may get strange build errors.
	Looks good! Try building with ./gradlew assembleDebug
	```
	**Workaround:** n/a This error is not accurate. OpenJDK packages the x86_64 and arm64 binaries together (see below). Question: do we want to standardize on which JDK we use? (needs investigation to see how m-c does it)

	```
	$ which java | xargs file

	/usr/bin/java: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64
	- Mach-O 64-bit executable x86_64] [arm64e:Mach-O 64-bit executable arm64e
	- Mach-O 64-bit executable arm64e]
	/usr/bin/java (for architecture x86_64):        Mach-O 64-bit executable x86_64
	/usr/bin/java (for architecture arm64e):        Mach-O 64-bit executable arm64e
	```

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-3319)

**Comment by skhamis** [comment]
Thanks for this @jonalmeida, we indeed need to revisit the build environment for M1 mac since when it was fixed there were a bunch of issues where aarch64 was not fully supported (I have an M1 Mac myself and try to fix most of the M1 problems that bubble up). I don't think that's the case anymore and we can make some better decisions about how people should onboard.

> Question: do we want to standardize on which JDK we use? (needs investigation to see how m-c does it)

That's an excellent idea, as it helps with the potential oddness from different JDKs and how they handle arm64 vs x86_64

I'll see if we can prioritize this a little since it's pretty brutal to get stuck on just trying to get the project to compile. Of course, we always welcome PRs if there are any quick wins we can get in right away.

**Comment by jonalmeida** [comment]
Returning back here, I realize that most of the issues in this list are not M1 specific, but changes to the different tools/SDKs that have happened over time so they can happen to anyone possibly.

> we always welcome PRs if there are any quick wins we can get in right away.

Yes, definitely! Will see what can be done to make it a bit smoother. :)

**Comment by titooan** [comment]
Hey! 
I went through the process of installing the system-level dependencies last week and here are the few issues I faced: 

OS: macOS Ventura 13.0
Hardware: MacBook Pro M1 Pro, 16-inch 2021

- The instructions tell to install python3. I first checked if it was already installed on my computer, it was, so I skipped this part. 
I discovered later that I needed to install it anyway using `brew`. 
It may be a good idea to emphasis that this step is needed even if python3 is already installed on the device, from an source other than `brew`.

- For some reason when trying to install gyp I followed the link provided on step b 
  > If you have additional questions, consult [this guide](https://github.com/mogemimi/pomdog/wiki/How-to-Install-GYP).

  I had troubles installing gyp following the instructions of this guide, and it worked better when I just followed the instructions of our [building.md](https://github.com/mozilla/application-services/blob/main/docs/building.md#building-application-services) doc. 


Basically the problems I faced were mainly caused by me not following exacty the instructions, but I wanted to document them here in case it happens to someone else.

**Comment by ncloudioj** [comment]
Would like to share some installation issues & workarounds from my recent experiences. (M1 w/o Resetta) 

- I needed to run `sudo xcode-select -s /Applications/Xcode.app/Contents/Developer`, otherwise `gyp` will raise exceptions from `xcode_emulation.py`
- I needed to manually set the `--target=arm64` in `build-nss-desktop.sh` when running `${NSS_SRC_DIR}/nss/build.sh`, otherwise the build for `nss` would fail at the linking stage. The doc of `build.sh` in `nss` indicates `--target=aarch64` can be used for overriding, but it turned out not working

**Comment by tarikeshaq** [comment]
There are a lot of great comments here, I moved the main issues into https://bugzilla.mozilla.org/show_bug.cgi?id=1889977
